src/System.Linq.Expressions/src/System/Runtime/CompilerServices/CallSite.cs

```diff
@@ -667,7 +667,7 @@ private T CreateCustomNoMatchDelegate(MethodInfo invoke)
                 Expression.Block(
                     Expression.Call(
                         typeof(CallSiteOps).GetMethod(nameof(CallSiteOps.SetNotMatched)),
-                        @params.First()
+                        @params[0]
```

Comment:
Note this changes the behavior of the following piece of code ever so slightly:
```csharp
CallSite<Foo>.Create(null);
...
delegate void Foo();
```
Instead of throwing `ArgumentOutOfRangeException` it now throws `IndexOutOfRangeException`. In fact, the case of a delegate without any parameters is not a valid configuration for `CallSite<T>`; the code uses a mix of various strategies to access the first-assumed-to-always-be-there parameter of the `Invoke` method, e.g. `IsSimpleSignature` uses array indexing as well. Given that compilers create `CallSite<T>` instances with delegates having at least one parameter, should we consider this a breaking change?